{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Time-resolved probing of RNase P\n",
    "\n",
    "This is a recreation of figures and analysis from Figure 3 of Ehrhardt & Weeks 2020.\n",
    "[pdf](https://weekslab.com/wp-content/uploads/sites/9/2021/01/2020_je_jacs.pdf)\n",
    "\n",
    "Raw sequencing reads, secondary structure diagram, and three-dimensional structure were obtained from the authors.\n",
    "Raw sequencing reads were analyzed using ShapeMapper v2.1.5 (with `--amplicon` and `--output-parsed-mutations` parameters).\n",
    "The resulting parsed mutation strings files (`_parsed.mut`) were analyzed further using RingMapper (default parameters).\n",
    "We focus on datasets at 0 (no Mg<sup>2+</sup>), 30 and 1200 seconds.\n",
    "There are thus eight required files for this analysis, included in this directory:\n",
    "\n",
    "- 3dhsCrystal_PlusLoops.pdb\n",
    "- RC_CRYSTAL_STRUCTURE.xrna\n",
    "- RC0_rnasep_profile.txt\n",
    "- RC0_rnasep_rings.txt\n",
    "- RC30_rnasep_profile.txt\n",
    "- RC30_rnasep_rings.txt\n",
    "- RC1200_rnasep_profile.txt\n",
    "- RC1200_rnasep_rings.txt\n",
    "\n",
    "## Import RNAvigate\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import rnavigate as rnav\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define the experimental samples and provide input file names\n",
    "\n",
    "Here we have 3 samples, TMO-treated RNase P construct:\n",
    "1. in the absence of Mg<sup>2+</sup> (No Mg<sup>2+</sup>)\n",
    "2. folded with Mg<sup>2+</sup> for 30 seconds (partially folded)\n",
    "3. folded with Mg<sup>2+</sup> for 1200 seconds (fully folded)\n",
    "\n",
    "Inputs common to all samples:\n",
    "- `ss`: a structure drawing file based on the crystal structure\n",
    "- `pdb`: 3D structure based on 3DHS as a .pdb file\n",
    "- One annotation for each of the regions defined in the original publication,\n",
    "  which are involved in the largest changes between time-points:\n",
    "  - `catalytic_core`, `P5`, `P2`, `loop_loop`\n",
    "\n",
    "Inputs unique to each sample:\n",
    "- `sample`: an arbitrary string used as a label on plots\n",
    "- `dmsmap`: a ShapeMapper2 profile.txt\n",
    "  - When using `dmsmap` instead of `shapemap` as the keyword, data are automatically rescaled using the same convention as PairMapper (manuscript citation 27).\n",
    "- `ringmap`: a RingMapper correlations file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_data = rnav.Sample(\n",
    "    sample='',\n",
    "    ss=\"RC_CRYSTAL_STRUCTURE.xrna\",\n",
    "    pdb={\"pdb\": \"3dhsCrystal_PlusLoops.pdb\",\n",
    "         \"chain\":\"A\"},\n",
    "    catalytic_core={\n",
    "        \"spans\": [[236, 240], [181, 188], [45, 47], [49, 49], [127, 128]],\n",
    "        \"sequence\": \"ss\",\n",
    "        \"color\": \"DarkOrchid\",\n",
    "        \"name\": \"catalytic core\"},\n",
    "    P5={\n",
    "        \"spans\": [[58, 62], [101, 105]],\n",
    "        \"sequence\": \"ss\",\n",
    "        \"color\": \"orange\",\n",
    "        \"name\": \"P5 helix\"},\n",
    "    P2={\n",
    "        \"spans\": [[189, 195], [15, 21]],\n",
    "        \"sequence\": \"ss\",\n",
    "        \"color\": \"DodgerBlue\",\n",
    "        \"name\": \"P2 helix\"},\n",
    "    loop_loop={\n",
    "        \"spans\": [[70, 75], [141, 146]],\n",
    "        \"sequence\": \"ss\",\n",
    "        \"color\": \"red\",\n",
    "        \"name\": \"L5.1 - L15.1\"},\n",
    "    )\n",
    "\n",
    "rc0 = rnav.Sample(\n",
    "    sample=\"No Mg$^{2+}$\",\n",
    "    inherit=common_data,\n",
    "    dmsmap=\"RC0_rnasep_profile.txt\",\n",
    "    ringmap=\"RC0_rnasep_rings.txt\")\n",
    "\n",
    "rc30 = rnav.Sample(\n",
    "    sample=\"partially folded\",\n",
    "    inherit=common_data,\n",
    "    dmsmap=\"RC30_rnasep_profile.txt\",\n",
    "    ringmap=\"RC30_rnasep_rings.txt\")\n",
    "\n",
    "rc1200 = rnav.Sample(\n",
    "    sample=\"fully folded\",\n",
    "    inherit=common_data,\n",
    "    dmsmap=\"RC1200_rnasep_profile.txt\",\n",
    "    ringmap=\"RC1200_rnasep_rings.txt\")\n",
    "\n",
    "sample_list = [rc0, rc30, rc1200]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Additional information for each sample\n",
    "\n",
    "The raw sequencing reads used in this analysis differ slightly from the original\n",
    "publication in that they are sequenced to much higher read depth. This higher\n",
    "read depth requires a slightly different filtering scheme. Instead of filtering\n",
    "for $G_{apc}$, which is dependent on read depth, we will filter using mutual\n",
    "information (MI) which is directly related to $G_{apc}$, but does not depend on\n",
    "read depth.\n",
    "\n",
    "$$ MI = \\frac{G_{apc}}{2 * depth} \\\\ $$\n",
    "\n",
    "Below, we calculate this value and add it to the \"ringmap\" dataframe for each\n",
    "sample, so that we can use it later to filter RINGs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample in sample_list:\n",
    "    sample.data[\"ringmap\"].data.eval(\"MI = 0.5 * Statistic / Mod_Depth\", inplace=True)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Custom color list\n",
    "\n",
    "We will use the annotations below to highlight important regions and also want\n",
    "to color individual nucleotides by region. We will make a custom list of\n",
    "colors, one color for each nucleotide in RNase P."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations_list = [\"catalytic_core\", \"P5\", \"P2\", \"loop_loop\"]\n",
    "color_list, cmap = common_data.get_data(\"ss\").get_colors_from_annotations(\n",
    "    annotations=common_data.get_data(annotations_list)\n",
    "    )\n",
    "\n",
    "plot = rnav.plots.ColorBar(1)\n",
    "plot.plot_data(cmap)\n",
    "plot.set_figure_size(width_ax_in=4)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display a detailed secondary structure drawing\n",
    "\n",
    "Here, we make a structure diagram that includes important details, as a\n",
    "reference for the more data-rich plots we will make later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = rnav.plot_ss(\n",
    "    samples=[rc0],\n",
    "    structure=\"ss\",                   # make a secondary structure diagram\n",
    "    colors={'sequence': color_list},  # use annotations colors\n",
    "    annotations=annotations_list,     # highlight our important regions\n",
    "    bp_style=\"conventional\",          # use Westoff-Leonitis base-pairing conventions\n",
    ")\n",
    "\n",
    "ax = plot.axes[0,0]\n",
    "# add these text labels to the plot axis\n",
    "# ax.text(x position, y position, \"displayed text\")\n",
    "ax.text(0.2, 1.5, \"catalytic\\ncore\", color=\"DarkOrchid\")\n",
    "ax.text(-6, 7.5, 'P1')\n",
    "ax.text(-6, -2.5, 'P4')\n",
    "ax.text(-6, -8.5, 'P5', color='orange')\n",
    "ax.text(-8.5, -8.5, 'P7')\n",
    "ax.text(0.1, -8.5, 'P15')\n",
    "ax.text(-6.5, -12.5, 'P5.1')\n",
    "ax.text(4.5, 15, 'L19')\n",
    "ax.text(8, 7.5, 'P19')\n",
    "ax.text(8, 2, 'P2', color='DodgerBlue')\n",
    "ax.text(8, -4.5, 'P3')\n",
    "ax.text(4.3, -5.5, 'P15.2')\n",
    "ax.text(4.5, -10.5, 'P15.1')\n",
    "ax.text(3, -17, 'L15.1', color='red')\n",
    "ax.text(-4.5, -17, 'L5.1', color='red');\n",
    "\n",
    "# plot.save(\"rnasep_big_ss.svg\")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display filtered RINGs on secondary structure drawing\n",
    "\n",
    "Next, for each of the samples, we plot the RING-MaP interactions in green on\n",
    "the secondary structure.\n",
    "\n",
    "Note that the RNA sequence used in the experimental data contains structure\n",
    "cassette sequences at the 5' and 3' ends, while the secondary structure and PDB\n",
    "structure do not. When RNAvigate plots data on a structure, a sequence\n",
    "alignment is performed to place the data in the correct position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_settings = {\n",
    "    'ss': {\n",
    "        'interactions': {\n",
    "            'linewidth': 8,\n",
    "            'alpha': 1},\n",
    "        'nucleotides': {\n",
    "            's': 6**2},\n",
    "        'structure': {\n",
    "            'zorder': 40,\n",
    "            'linewidth': 6},\n",
    "        'basepairs': {\n",
    "            'zorder': 39}\n",
    "    }}\n",
    "with rnav.styles.Settings(temp_settings):\n",
    "    plot = rnav.plot_ss(\n",
    "        samples=sample_list,           # Make one structure diagram for each sample in our list\n",
    "        structure=\"ss\",\n",
    "        interactions={\n",
    "            'interactions': \"ringmap\", # show ringmap interactions\n",
    "            'structure': 'ss',         # use the \"ss\" structure drawing to filter by structural features\n",
    "            'MI_ge': 0.00025,          # Only include RINGs with MI value greater than or equal to 0.00025\n",
    "            'Statistic_ge':23,         # Only include RINGs with Gapc value greater than or equal to 23\n",
    "            'min_cd': 6,               # Only include RINGs with contact distance of at least 5\n",
    "            'positive_only': True,     # Only include RINGs with a positive correlation\n",
    "            'cmap':'limegreen',         # Set the color of all rings to \"limegreen\"\n",
    "            },\n",
    "        bp_style=\"line\",               # represent base-pairs with solid lines\n",
    "        colors={\n",
    "            'structure': color_list,   # Apply custom color list to the backbone of the structure\n",
    "            'nucleotides': color_list,\n",
    "            },\n",
    "        )\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Color RINGs from previous plot by origination time-point\n",
    "\n",
    "The above plot is nice, but we would like to convey additional information.\n",
    "Certain RINGs are newly formed at each time point, and others are carried over\n",
    "from a previous time point. We will color these differently. The code below\n",
    "includes some advanced manipulations of the underlying dataframes to acheive\n",
    "this effect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# retreive the RING-MaP DataFrame for each sample\n",
    "s1 = rc0.data['ringmap'].data\n",
    "s2 = rc30.data['ringmap'].data\n",
    "s3 = rc1200.data['ringmap'].data\n",
    "# get i_offset and j_offset values as 2D numpy array\n",
    "a = s1[['i', 'j']].values\n",
    "b = s2[['i', 'j']].values\n",
    "c = s3[['i', 'j']].values\n",
    "# determine which rings are newly formed and store result in the dataframe\n",
    "# is there 'any' row in a that completely matches all values of a row in 'b'\n",
    "s1['new_rings'] = s1['mask']\n",
    "s2['new_rings'] = ~(b[:, None] == a[s1['mask']]).all(-1).any(-1)\n",
    "s3['new_rings'] = ~(c[:, None] == b[s2['mask']]).all(-1).any(-1)\n",
    "\n",
    "with rnav.styles.Settings(temp_settings):\n",
    "    plot = rnav.plot_ss(\n",
    "        samples=sample_list,\n",
    "        structure=\"ss\",\n",
    "        interactions={\n",
    "            'interactions': 'ringmap',\n",
    "            'structure': 'ss',\n",
    "            'MI_ge': 0.00025,\n",
    "            'Statistic_ge':23,\n",
    "            'min_cd': 6,\n",
    "            'positive_only': True,\n",
    "            'cmap':['khaki', 'limegreen'],\n",
    "            'metric': 'new_rings',\n",
    "            'normalization': 'none'},\n",
    "        bp_style='line',\n",
    "        colors={\n",
    "            'structure': color_list,\n",
    "            'nucleotides': color_list},\n",
    "        colorbars=False\n",
    "        )\n",
    "\n",
    "# add text labels for colored rings\n",
    "plot.axes[0, 0].text(4, 8, 'newly formed\\nin current\\ntime point', color='limegreen', fontsize=16, ha='right')\n",
    "plot.axes[0, 1].text(4, 8, 'present in\\nprevious\\ntime point', color='khaki', fontsize=16, ha='right');\n",
    "\n",
    "# save plot to file\n",
    "# plot.save('rnasep_rings_ss.svg')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display RINGs density and regions of interest using heatmaps\n",
    "\n",
    "Heatmaps are a good way to view very dense 2D data but can be hard to\n",
    "interpret. The reader is helped here by highlighting regions which\n",
    "illustrate time-dependent structural interactions.\n",
    "\n",
    "Here, we will use the kernal density estimate (\"kde\") to simplify the\n",
    "visualization, an improvement over the published figure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "regions = [((225,250), (80,110)),     # ((x1, x2), (y1, y2))\n",
    "           ((225,255), (175,200)),    # highlights interactions between region\n",
    "           ((135,150), (60,85)),      # x1-x2 and y1-y2\n",
    "           ((60,155), (10, 30))]\n",
    "\n",
    "plot = rnav.plot_heatmap(\n",
    "    samples=sample_list,              # 1 plot for each of three samples\n",
    "    sequence=\"ss\",                    # align all data to the secondary structure sequence\n",
    "    structure='ss',                   # use nucleotide positions from 'ss' data (excludes the structure cassetes)\n",
    "    interactions={\n",
    "        'interactions': 'ringmap',    # Plot RING-MaP interactions\n",
    "        'positive_only': True,        # remove negative correlations\n",
    "        'cmap': 'Greens',             # use matplotlib 'Greens' colormap\n",
    "        },\n",
    "    regions=regions,                  # draw boxes around regions\n",
    "    plot_type='kde')                  # plot RINGs density instead of individual RINGs\n",
    "\n",
    "# adding labels to region boxes\n",
    "ax = plot.axes[0,0]\n",
    "arrowprops = {'arrowstyle':'-', 'color': 'black'}\n",
    "ax.annotate('P4-P7', [75, 205])\n",
    "ax.annotate('P4-P19', [187.5, 220], [200, 175], arrowprops=arrowprops)\n",
    "ax.annotate('L5.1-L15.1', [90, 142.5], [140, 120], arrowprops=arrowprops)\n",
    "ax.annotate('P2-P5,P7', [35, 75], [75, 50], arrowprops=arrowprops)\n",
    "\n",
    "# adding labels for density estimation\n",
    "ax = plot.axes[0, 1]\n",
    "ax.annotate('more\\ndense', [170, 175], [200, 150], arrowprops=arrowprops)\n",
    "ax.annotate('less\\ndense', [125, 130], [150, 100], arrowprops=arrowprops)\n",
    "\n",
    "# save plots to files\n",
    "# plot.save('rnasep_heatmaps.svg')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display filtered RINGs on 3D structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = rnav.plot_mol(\n",
    "    samples=[rc1200],\n",
    "    structure=\"pdb\",\n",
    "    interactions={\n",
    "        'interactions': 'ringmap',\n",
    "        'structure': 'ss',\n",
    "        'MI_ge': 0.000025,\n",
    "        'Zij_ge':2,\n",
    "        'ss_only': True,\n",
    "        'Statistic_ge':23,\n",
    "        'min_cd': 6,\n",
    "        'positive_only': True,\n",
    "        'cmap': 'limegreen'},\n",
    "    colors=color_list,\n",
    "    hide_cylinders=True,\n",
    "    width=800,\n",
    "    height=800,\n",
    "    background_alpha=0,\n",
    "    orientation=[-35.06,-98.79,-28.57,-245.66,0.14,0.52,0.62,-0.57],\n",
    "    title=False\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot.save()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.2 ('RNAvigate')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "5cbb00885ecd6c28cb92c1f604a30a23977756562934ed05197ce2fdb8002d56"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
